Add trace batching forward/backward rule

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49979 Add trace batching forward/backward rule**

Differential Revision: [D25734379](https://our.internmc.facebook.com/intern/diff/D25734379)

ejguan committed:
Add trace batching forward/backward rule

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -287,6 +287,30 @@ Tensor squeeze_dim_batching_rule(const Tensor& self, int64_t dim) {
+Tensor trace_batching_rule(const Tensor& self) {
+  auto self_physical = MultiBatchVmapTransform::logicalToPhysical(self);
+  // Batched Diagonal View
+  auto self_diag = at::diagonal(self_physical.tensor(), 0, -2, -1);
```
nit: comments can improve readability of arguments:
```
at::diagonal(self_physical.tensor(), /*offset*/0, /*dim1*/-2, /*dim2*/-1);
```

zou3519 commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -287,6 +287,30 @@ Tensor squeeze_dim_batching_rule(const Tensor& self, int64_t dim) {
+  auto grad_input = at::zeros(grad_physical.getPhysicalShape(input_sizes), grad.options());
+  // Batched Diagonal View
+  auto grad_input_diag = at::diagonal(grad_input, 0, -2, -1);
+  // Reshape grad output
+  auto grad_physical_size = grad_physical.tensor().sizes();
+  VmapDimVector view_shape(grad_physical.numBatchDims() + 1, 1);
+  std::copy(grad_physical_size.begin(),
+      grad_physical_size.begin() + grad_physical.numBatchDims(),
+      view_shape.begin());
+  auto grad_physical_tensor = grad_physical.tensor().view(view_shape);
```
Is this the same thing as doing `grad_physical.tensor().unsqueeze(-1)`?

ejguan committed:
Update on "Add trace batching forward/backward rule"


Differential Revision: [D25734379](https://our.internmc.facebook.com/intern/diff/D25734379)

[ghstack-poisoned]
